### PR TITLE
rpl: omit dodag id in DAO(-ACK) for global instances

### PR DESCRIPTION
**Pull request by cgundogan**
This PR introduces minor support for global / local instance ids.
If an instance id `0 <= id <= 127` is specified, the instance is global and thus the `dodag_id` will be omitted in outgoing `DAOs` and `DAO-ACKs` and not expected respectively on the receiving side. (see https://github.com/RIOT-OS/RIOT/pull/3050#discussion_r36549990)

### CODE REVIEW

**Comment by gebart** [comment]
needs rebase :/

File: sys/net/routing/ng_rpl/ng_rpl_control_messages.c [hunk 1]
```diff
@@ -529,18 +538,23 @@ void ng_rpl_send_DAO(ng_rpl_dodag_t *dodag, ipv6_addr_t *destination, uint8_t li
...
-    /* set the D flag to indicate that a DODAG id is present */
+    if (local_instance) {
+        /* set the D flag to indicate that a DODAG id is present */
+        dao->k_d_flags = (1 << 6);
+        memcpy((dao + 1), &dodag->dodag_id, sizeof(ipv6_addr_t));
+        target = (ng_rpl_opt_target_t *)(((uint8_t *) target) + sizeof(ipv6_addr_t));
+    }
     /* set the K flag to indicate that a ACKs are required */
-    dao->k_d_flags = ((1 << 6) | (1 << 7));
+    dao->k_d_flags |= (1 << 7);
```
Comment (gebart) [inline sys/net/routing/ng_rpl/ng_rpl_control_messages.c]: in the case of `!local_instance` the k_d_flags member is never zeroed out.

File: sys/net/routing/ng_rpl/ng_rpl_control_messages.c [hunk 2]
```diff
@@ -582,11 +601,13 @@ void ng_rpl_send_DAO_ACK(ng_rpl_dodag_t *dodag, ipv6_addr_t *destination, uint8_
     dao_ack = (ng_rpl_dao_ack_t *)(icmp + 1);
 
     dao_ack->instance_id = dodag->instance->id;
-    /* set the D flag to indicate that a DODAG id is present */
-    dao_ack->d_reserved = (1 << 7);
+    if (local_instance) {
+        /* set the D flag to indicate that a DODAG id is present */
+        dao_ack->d_reserved = (1 << 7);
+        memcpy((dao_ack + 1), &dodag->dodag_id, sizeof(ipv6_addr_t));
+    }
```
Comment (gebart) [inline sys/net/routing/ng_rpl/ng_rpl_control_messages.c]: Add an `else` statement which sets `dao_ack->d_reserved` to zero

File: sys/net/routing/ng_rpl/ng_rpl_control_messages.c [hunk 3]
```diff
@@ -595,25 +616,34 @@ void ng_rpl_recv_DAO(ng_rpl_dao_t *dao, ipv6_addr_t *src, uint16_t len)
...
         return;
     }
 
+    len -= (sizeof(ng_rpl_dao_t) + sizeof(icmpv6_hdr_t));
+
     /* check if the D flag is set before accessing the DODAG id */
-    if (!(dao->k_d_flags & (1 << 6))) {
-        DEBUG("RPL: DAO with D flag unset - global instances not supported\n");
-        return;
+    if ((dao->k_d_flags & (1 << 6))) {
```
Comment (gebart) [inline sys/net/routing/ng_rpl/ng_rpl_control_messages.c]: please replace `(1 << 6)` with a new constant macro with a describing name

File: sys/net/routing/ng_rpl/ng_rpl_control_messages.c [hunk 4]
```diff
@@ -582,11 +601,13 @@ void ng_rpl_send_DAO_ACK(ng_rpl_dodag_t *dodag, ipv6_addr_t *destination, uint8_
     dao_ack = (ng_rpl_dao_ack_t *)(icmp + 1);
 
     dao_ack->instance_id = dodag->instance->id;
-    /* set the D flag to indicate that a DODAG id is present */
-    dao_ack->d_reserved = (1 << 7);
+    if (local_instance) {
+        /* set the D flag to indicate that a DODAG id is present */
+        dao_ack->d_reserved = (1 << 7);
```
Comment (gebart) [inline sys/net/routing/ng_rpl/ng_rpl_control_messages.c]: please replace `(1 << 7)` with a new constant macro with a describing name

File: sys/net/routing/ng_rpl/ng_rpl_control_messages.c (hunk 1 above)
Comment (gebart) [inline sys/net/routing/ng_rpl/ng_rpl_control_messages.c]: please replace `(1 << 7)` with a new constant macro with a describing name

File: sys/net/routing/ng_rpl/ng_rpl_control_messages.c [hunk 5]
```diff
@@ -529,18 +538,23 @@ void ng_rpl_send_DAO(ng_rpl_dodag_t *dodag, ipv6_addr_t *destination, uint8_t li
 
     icmp = (icmpv6_hdr_t *)pkt->data;
     dao = (ng_rpl_dao_t *)(icmp + 1);
+    target = (ng_rpl_opt_target_t *) (dao + 1);
 
     dao->instance_id = dodag->instance->id;
-    /* set the D flag to indicate that a DODAG id is present */
+    if (local_instance) {
+        /* set the D flag to indicate that a DODAG id is present */
+        dao->k_d_flags = (1 << 6);
```
Comment (gebart) [inline sys/net/routing/ng_rpl/ng_rpl_control_messages.c]: please replace `(1 << 6)` with a new constant macro with a describing name

File: sys/net/routing/ng_rpl/ng_rpl_control_messages.c [hunk 6]
```diff
@@ -595,25 +616,34 @@ void ng_rpl_recv_DAO(ng_rpl_dao_t *dao, ipv6_addr_t *src, uint16_t len)
...
     }
 
+    len -= (sizeof(ng_rpl_dao_t) + sizeof(icmpv6_hdr_t));
+
     /* check if the D flag is set before accessing the DODAG id */
-    if (!(dao->k_d_flags & (1 << 6))) {
-        DEBUG("RPL: DAO with D flag unset - global instances not supported\n");
-        return;
+    if ((dao->k_d_flags & (1 << 6))) {
+        if ((dodag = ng_rpl_dodag_get(inst, (ipv6_addr_t *)(dao + 1))) == NULL) {
```
Comment (gebart) [inline sys/net/routing/ng_rpl/ng_rpl_control_messages.c]: `len` must be checked `>sizeof(ipv6_addr_t)` before dereferencing `dao+1`

File: sys/net/routing/ng_rpl/ng_rpl_control_messages.c [hunk 7]
```diff
@@ -595,25 +616,34 @@ void ng_rpl_recv_DAO(ng_rpl_dao_t *dao, ipv6_addr_t *src, uint16_t len)
...
     ng_rpl_instance_t *inst = NULL;
     ng_rpl_dodag_t *dodag = NULL;
+    ng_rpl_opt_t *opts = (ng_rpl_opt_t *) (dao + 1);
+
     if ((inst = ng_rpl_instance_get(dao->instance_id)) == NULL) {
         DEBUG("RPL: DAO with unknown instance id (%d) received\n", dao->instance_id);
         return;
     }
 
+    len -= (sizeof(ng_rpl_dao_t) + sizeof(icmpv6_hdr_t));
```
Comment (gebart) [inline sys/net/routing/ng_rpl/ng_rpl_control_messages.c]: check `len` to be big enough first

Comment (gebart) [inline sys/net/routing/ng_rpl/ng_rpl_control_messages.c]: Actually, none of the receive functions check the length of the packets before processing them. This was an oversight when reviewing #3050 and needs to be fixed, but it can be done in a separate PR if you prefer to keep this PR more to the point.

**Comment by gebart** [comment]
Wow! :clap: 

This is one more step towards Contiki interoperability

**Comment by cgundogan** [comment]
@gebart, addressed your comments and did a rebase. I will put the length checks in a separate PR however, thanks again for the reminder.

**Comment by gebart** [comment]
OK to squash.
Untested ACK. I won't have the possibility to test this PR during this week, anyone else feel like it?

**Comment by OlegHahm** [comment]
I can test tomorrow. Please send me a reminder.

**Comment by cgundogan** [comment]
squashed. anybody can test this by starting rpl with an instance id in the range of `1 <= id <= 127` and look at the `DAO`s in wireshark. Then do the same with a rpl instance with id `>= 128` and look again at the `DAO`s. In one case there should be the `DODAG-ID` present, for the other case not.

File: sys/net/gnrc/routing/rpl/gnrc_rpl_control_messages.c [hunk 8]
```diff
@@ -28,17 +28,20 @@
 #define ENABLE_DEBUG    (0)
 #include "debug.h"
 
-#if ENABLE_DEBUG && defined(MODULE_IPV6_ADDR)
```
Comment (OlegHahm) [inline sys/net/gnrc/routing/rpl/gnrc_rpl_control_messages.c]: How is this related to the PR?

File: sys/net/gnrc/routing/rpl/gnrc_rpl_control_messages.c [hunk 9]
```diff
@@ -640,15 +681,19 @@ void gnrc_rpl_recv_DAO_ACK(gnrc_rpl_dao_ack_t *dao_ack)
...
+        }
     }
-
-    if ((dodag = gnrc_rpl_dodag_get(inst, &dao_ack->dodag_id)) == NULL) {
-        DEBUG("RPL: DAO-ACK with unknown DODAG id (%s)\n", ipv6_addr_to_str(addr_str,
-                    &dao_ack->dodag_id, sizeof(addr_str)));
-        return;
+    /* D flag not set - global instance id */
+    else {
+        if ((dodag = gnrc_rpl_dodag_get(inst, NULL)) == NULL) {
```
Comment (OlegHahm) [inline sys/net/gnrc/routing/rpl/gnrc_rpl_control_messages.c]: else if {

File: sys/net/gnrc/routing/rpl/gnrc_rpl_control_messages.c (hunk 8 above)
Comment (cgundogan) [inline sys/net/gnrc/routing/rpl/gnrc_rpl_control_messages.c]: unrelated, but also unnecessary as rpl does not build without `MODULE_IPV6_ADDR`. Do you want me to open a separate PR for this change?

Comment (OlegHahm) [inline sys/net/gnrc/routing/rpl/gnrc_rpl_control_messages.c]: No, but a separate commit would be great.

**Comment by cgundogan** [comment]
addressed @OlegHahm's comments (squashed)

File: sys/net/gnrc/routing/rpl/gnrc_rpl_control_messages.c (hunk 8 above)
Comment (cgundogan) [inline sys/net/gnrc/routing/rpl/gnrc_rpl_control_messages.c]: this comment has been addressed

**Comment by cgundogan** [comment]
rebased to curr. master

**Comment by cgundogan** [comment]
>I can test tomorrow. Please send me a reminder.

Friendly reminder, @OlegHahm

**Comment by OlegHahm** [comment]
> If an instance id 0 <= id <= 127

```
1440077466.212608;m3-9;> rpl root 0 affe::1
1440077466.213585;m3-9;<instance_id> must be a positive number greater than zero
```

**Comment by OlegHahm** [comment]
Seems not to break anything and Wireshark shows what is described. ACK
